src/mscorlib/src/System/String.Comparison.cs

```diff
@@ -817,153 +817,106 @@ public override bool Equals(Object obj)
 
-            switch (comparisonType) {
-                case StringComparison.CurrentCulture:
-                    return (CultureInfo.CurrentCulture.CompareInfo.Compare(this, value, CompareOptions.None) == 0);
-
-                case StringComparison.CurrentCultureIgnoreCase:
-                    return (CultureInfo.CurrentCulture.CompareInfo.Compare(this, value, CompareOptions.IgnoreCase) == 0);
-
-                case StringComparison.InvariantCulture:
-                    return (CultureInfo.InvariantCulture.CompareInfo.Compare(this, value, CompareOptions.None) == 0);
-
-                case StringComparison.InvariantCultureIgnoreCase:
-                    return (CultureInfo.InvariantCulture.CompareInfo.Compare(this, value, CompareOptions.IgnoreCase) == 0);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool EqualsOrdinalCaseSensitive(String a, String b) {
+            if (ReferenceEquals(a, b)) return true;
+            if (NotPossiblyEquals(a, b)) return false;
 
-                case StringComparison.Ordinal:
-                    if (this.Length != value.Length)
-                        return false;
-                    return EqualsHelper(this, value);
+            return EqualsHelper(a, b);
+        }
 
-                case StringComparison.OrdinalIgnoreCase:
-                    if (this.Length != value.Length)
-                        return false;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool EqualsOrdinalIgnoreCase(String a, String b) {
+            if (ReferenceEquals(a, b)) return true;
+            if (NotPossiblyEquals(a, b)) return false;
 
-                    // If both strings are ASCII strings, we can take the fast path.
-                    if (this.IsAscii() && value.IsAscii()) {
-                        return (CompareOrdinalIgnoreCaseHelper(this, value) == 0);
-                    }
+            var isAscii = a.IsAscii();
+            if (isAscii != b.IsAscii()) return false;
+            if (isAscii)
+            {
+                return (CompareOrdinalIgnoreCaseHelper(a, b) == 0);
+            }
+            // Take the slow path.
 
 #if FEATURE_COREFX_GLOBALIZATION
-                    return (CompareInfo.CompareOrdinalIgnoreCase(this, 0, this.Length, value, 0, value.Length) == 0);
+            return (CompareInfo.CompareOrdinalIgnoreCase(a, 0, a.Length, b, 0, b.Length) == 0);
 #else
-                    // Take the slow path.
-                    return (TextInfo.CompareOrdinalIgnoreCase(this, value) == 0);
+            return (TextInfo.CompareOrdinalIgnoreCase(a, b) == 0);
 #endif
-
-                default:
-                    throw new ArgumentException(Environment.GetResourceString("NotSupported_StringComparison"), nameof(comparisonType));
-            }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool EqualsCulture(String a, String b, StringComparison comparisonType) {
```

Comment:
it is quite a big method to inline, is culture comparison common enough to motivate that?